### PR TITLE
Nats

### PR DESCRIPTION
**Pull request by BohdanRysakov**

### CODE REVIEW

File: nats/src/main/kotlin/rys/nats/controller/NatsChatFindOneController.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                    successBuilder.apply {
+                        this.result = Mongochat.Chat.newBuilder().apply {
+                            id = chat!!.id.toString()
+                            name = chat.name
+                            chat.users.forEach {
+                                this.addUsers(it.toString())
+                            }
+                        }.build()
+                    }
+                }.build()
```
Comment (HNDaG) [inline nats/src/main/kotlin/rys/nats/controller/NatsChatFindOneController.kt:40]: Not completely sure, but doesn't you call **newBuilder()** tree times? So why you call **build()** only twice? In other controllers it has similar structure.

**Comment by BohdanRysakov** [comment]
@ajax-gnatiuk-s  
Q: 
1) I have additional layer for serialization (nats/src/main/kotlin/rys/nats/utils/NatsValidMongoChatParser.kt ) 
I can't say why for now I need it but I feel like this layer must be. For example: after update of Documents & Protos we may add logic in order to add fields in serialisation. 
2) Maybe proto folder need to be in other place? 

Also don't mind files from another modules - only nats package need to be reviewed.

File: nats/src/main/kotlin/rys/nats/controller/NatsChatUpdateController.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                        this.result = Mongochat.Chat.newBuilder().apply {
+                            id = updatedChat.id.toString()
+                            name = updatedChat.name
+                            updatedChat.users.forEach {
+                                this.addUsers(it.toString())
+                            }
+                        }.build()
+                    }
+                }.build()
+
```
Comment (HNDaG) [inline nats/src/main/kotlin/rys/nats/controller/NatsChatUpdateController.kt:46]: Don't you think that method is too complex? Why not to separete building successful and failed response?

File: nats/src/test/kotlin/rys/nats/NatsTestConfiguration.kt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    @Bean
+    fun mongoClient(): MongoClient {
+        return MongoClients.create(mongoUri)
+    }
+
+    @Bean
+    fun mongoTemplate(@Autowired mongoClient: MongoClient): MongoTemplate {
+        return MongoTemplate(mongoClient, "testDB")
+    }
```
Comment (HNDaG) [inline nats/src/test/kotlin/rys/nats/NatsTestConfiguration.kt:42]: Do you really need this bean configurations? As I know **MongoTemplate** is configured automatically with some starters. And as for **MongoClient**, I failed to find where you use it. I think that db connect are configured via properties file.

File: rest/gradlew
Comment (HNDaG) [inline rest/gradlew:1]: Consider having all gradle files and wrapper in one instance, in one top level module.

File: build.gradle.kts
Comment (HNDaG) [inline build.gradle.kts:1]: Consider refactoring your settings and build gradle files. As for now it's looks for me as 2 separate projects that you just include not to another via dependency.

Comment (HNDaG) [inline build.gradle.kts:1]: Consider, making 2 modules. One for proto file,  Nats subject and future Kafka topic. And second one with your code logic. Or if refactoring your project even 3 modules(maybe not). (Inner Api Convention (v2)).

**Review by HNDaG — COMMENTED** [review]
Take it on your own risk)

File: rest/src/main/kotlin/rys/rest/service/impl/UserServiceImplementation.kt [hunk 4]
```diff
@@ -1,12 +1,12 @@
-package rys.ajaxpetproject.service.impl
+package rys.rest.service.impl
```
Comment (ajax-gnatiuk-s) [inline rest/src/main/kotlin/rys/rest/service/impl/UserServiceImplementation.kt:1]: Mostly, a project has a basic package like: `systems.ajax`, `com.google`, `com.github`

Ok, you have chosen `rys` as root. Then choose the next basic subdomain, for example `ajaxpetproject` whatever. Just choose the application name.

so, your basic, mandatory and common package for all project would be something like `rys.ajaxpetproject`.

Then for the `rest` basic package would be `rys.ajaxpetproject.rest`, for nats `rys.ajaxpetproject.nats`, etc.

Don't make your task harder

File: rest/src/main/kotlin/rys/rest/service/impl/ChatServiceImpl.kt [hunk 5]
```diff
@@ -1,35 +1,39 @@
...
 
-    override fun updateChat(id: ObjectId, updatedMongoChat: MongoChat): MongoChat =
+    override fun updateChat(id: ObjectId, updatedMongoChat: MongoChat): MongoChat {
         findChatById(id)
-            .let { chatRepository.save(updatedMongoChat) }
+            .let {
+                val newChat = updatedMongoChat.copy(id = id)
+                chatRepository.save(newChat)
+            }
+        return  findChatById(id)!!
```
Comment (ajax-gnatiuk-s) [inline rest/src/main/kotlin/rys/rest/service/impl/ChatServiceImpl.kt:28]: it's not clear why:
1. you invoke findChatById twice
2. why you have changed the function body at all

File: rest/src/main/kotlin/rys/rest/exceptions/ChatNotFoundException.kt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package rys.rest.exceptions
+
+class ChatNotFoundException : RuntimeException {
+    constructor(message: String?) : super(message)
+    constructor(message: String?, cause: Throwable?) : super(message, cause)
+}
```
Comment (ajax-gnatiuk-s) [inline rest/src/main/kotlin/rys/rest/exceptions/ChatNotFoundException.kt:6]:
```suggestion
class ChatNotFoundException(message: String) : RuntimeException(message)
```

maybe this variant covers all your cases?

File: rest/src/main/kotlin/rys/rest/RestApplication.kt [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+import org.springframework.boot.autoconfigure.SpringBootApplication
+import org.springframework.boot.runApplication
+
+@SpringBootApplication
+class RestApplication
+
+@Suppress("SpreadOperator")
+fun main(args: Array<String>) {
+    runApplication<RestApplication>(*args)
+}
```
Comment (ajax-gnatiuk-s) [inline rest/src/main/kotlin/rys/rest/RestApplication.kt:12]: why you need the additional `main` in the project?

File: rest/gradlew.bat [hunk 8]
```diff
@@ -0,0 +1,92 @@
+@rem
```
Comment (ajax-gnatiuk-s) [inline rest/gradlew.bat:1]: need discuss this

File: nats/src/main/proto/mongochat.proto [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+  message Failure {
+    string message = 1;
+
+    oneof error {
+      Error internal_error = 2;
+
+      Error not_found = 3;
+    }
+  }
+}
```
Comment (ajax-gnatiuk-s) [inline nats/src/main/proto/mongochat.proto:144]: :+1:

File: nats/src/main/kotlin/rys/nats/utils/NatsValidMongoChatParser.kt [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+import rys.nats.protostest.Mongochat.ChatDeleteRequest
+import rys.nats.protostest.Mongochat.ChatDeleteResponse
+import rys.nats.protostest.Mongochat.ChatFindAllResponse
+import rys.nats.protostest.Mongochat.ChatFindOneRequest
+import rys.nats.protostest.Mongochat.ChatFindOneResponse
+import rys.nats.protostest.Mongochat.ChatUpdateRequest
+import rys.nats.protostest.Mongochat.ChatUpdateResponse
+
+@Suppress("TooManyFunctions", "NestedBlockDepth")
+object NatsValidMongoChatParser {
```
Comment (ajax-gnatiuk-s) [inline nats/src/main/kotlin/rys/nats/utils/NatsValidMongoChatParser.kt:14]: we don't need this class. remove it

File: nats/src/main/kotlin/rys/nats/exception/InternalException.kt [hunk 11]
```diff
@@ -0,0 +1,7 @@
+package rys.nats.exception
+
+class InternalException : RuntimeException {
+    constructor() : super("Parsing error. See logs for further details.")
+    constructor(message: String?) : super(message)
+    constructor(message: String?, cause: Throwable?) : super(message, cause)
+}
```
Comment (ajax-gnatiuk-s) [inline nats/src/main/kotlin/rys/nats/exception/InternalException.kt:7]:
```suggestion
class InternalException(message: String) : RuntimeException(message)
```

Comment (ajax-gnatiuk-s) [reply nats/src/main/kotlin/rys/nats/exception/InternalException.kt:7]: I believe this approach will cover all your cases

File: nats/src/main/kotlin/rys/nats/config/NatsConfig.kt [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+
+@Configuration
+class NatsConfig {
+    @Value("\${nats.uri}")
+    private lateinit var natsUri: String
+
+    private val log = LoggerFactory.getLogger(this::class.java)
```
Comment (ajax-gnatiuk-s) [inline nats/src/main/kotlin/rys/nats/config/NatsConfig.kt:16]:
```suggestion
    companion object {
        private val log = LoggerFactory.getLogger(NatsConfig::class.java)
    }
```

**Comment by ajax-gnatiuk-s** [comment]
I've set up 1-2-1 tomorrow to discuss your PR

File: src/main/kotlin/rys/ajaxpetproject/AjaxPetProject.kt [hunk 13]
```diff
@@ -3,10 +3,10 @@ package rys.ajaxpetproject
 import org.springframework.boot.autoconfigure.SpringBootApplication
 import org.springframework.boot.runApplication
 
-@SpringBootApplication
-class AjaxPetProjectApplication
+@SpringBootApplication(scanBasePackages = ["rys"])
```
[the remark this answers is not in the document]
Comment (BohdanRysakov) [reply src/main/kotlin/rys/ajaxpetproject/AjaxPetProject.kt:6]: Spring didn't manually manage another modules. (For me that's also strange) 
So it works like :
rys.ajaxpetprojext (Main here )
       -rys.nats
       -rys.rest
      
And rest's and ants's beans didn't created by default

**Pull request merged**